java/query/builder/Sortable.java

```diff
@@ -21,42 +21,48 @@
 
 package com.vaticle.typeql.lang.query.builder;
 
+import com.vaticle.typedb.common.collection.Pair;
 import com.vaticle.typeql.lang.common.TypeQLArg;
+import com.vaticle.typeql.lang.common.exception.ErrorMessage;
 import com.vaticle.typeql.lang.common.exception.TypeQLException;
 import com.vaticle.typeql.lang.pattern.variable.UnboundVariable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
-import java.util.stream.Collectors;
+
 import static com.vaticle.typeql.lang.common.TypeQLToken.Char.COMMA_SPACE;
 import static com.vaticle.typeql.lang.common.TypeQLToken.Char.SPACE;
-import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_SORTING_ORDER;
-import static java.util.stream.Stream.concat;
-import static java.util.stream.Stream.of;
+import static com.vaticle.typeql.lang.common.exception.ErrorMessage.VARIABLE_NOT_SORTED;
 
 public interface Sortable<S, O, L> {
 
     default S sort(String var, String... vars) {
-        return sort(concat(of(var), of(vars)).map(UnboundVariable::named).collect(Collectors.toList()));
+        List<Pair<UnboundVariable, TypeQLArg.Order>> pairs = new ArrayList<>();
+        pairs.add(new Pair<>(UnboundVariable.named(var), null));
+        for (String v : vars) pairs.add(new Pair<>(UnboundVariable.named(v), null));
+        return sort(pairs);
     }
 
-    default S sort(UnboundVariable var, UnboundVariable... vars) {
-        return sort(concat(of(var), of(vars)).collect(Collectors.toList()));
+    default S sort(Pair<String, String> varOrder) {
+        return sort(varOrder, Collections.emptyList());
     }
 
-    default S sort(List<String> vars, String order) {
-        TypeQLArg.Order o = TypeQLArg.Order.of(order);
-        if (o == null) throw TypeQLException.of(
-                INVALID_SORTING_ORDER.message(TypeQLArg.Order.ASC, TypeQLArg.Order.DESC)
-        );
-        return sort(vars.stream().map(UnboundVariable::named).collect(Collectors.toList()), o);
-    }
-
-    default S sort(List<UnboundVariable> vars) {
-        return sort(new Sorting(vars));
+    default S sort(Pair<String, String> varOrder, List<Pair<String, String>> varOrders) {
```

Comment:
the secondary "shorthand" entry point is using pairs of strings, eg `pair("x", "asc")` to indicate a variable's order.

However, this clashes with the canonical API (below), so we split it into the two methods above, the second of which takes a list of following variable orderings